### PR TITLE
feat(new_metrics): implement the counter

### PR DESCRIPTION
**Pull request by empiredan**
https://github.com/apache/incubator-pegasus/issues/931

### CODE REVIEW

File: include/dsn/utility/metrics.h [hunk 1]
```diff
@@ -68,11 +69,26 @@
...
     ::dsn::gauge_prototype<double> METRIC_##name({#entity_type, #name, unit, desc, ##__VA_ARGS__})
+// There are 2 kinds of counters. `counter` is the general type of counter that are implemented
+// by striped_long_adder, which can achieve high performance while consuming less memory if
+// it's not updated very frequently. `counter2` uses concurrent_long_adder as the underlying
+// implementation. It has higher performance while consuming more memory if it's updated very
+// frequently. See also include/dsn/utility/long_adder.h for details.
+#define METRIC_DEFINE_counter(entity_type, name, unit, desc, ...)                                  \
+    ::dsn::counter_prototype<::dsn::striped_long_adder> METRIC_##name(                             \
+        {#entity_type, #name, unit, desc, ##__VA_ARGS__})
+#define METRIC_DEFINE_counter2(entity_type, name, unit, desc, ...)                                 \
```
Comment (acelyc111) [inline include/dsn/utility/metrics.h:80]: `counter2` is not a good naming, better to use a more meaningful name.

File: include/dsn/utility/metrics.h [hunk 2]
```diff
@@ -68,11 +69,26 @@
     ::dsn::gauge_prototype<int64_t> METRIC_##name({#entity_type, #name, unit, desc, ##__VA_ARGS__})
 #define METRIC_DEFINE_gauge_double(entity_type, name, unit, desc, ...)                             \
     ::dsn::gauge_prototype<double> METRIC_##name({#entity_type, #name, unit, desc, ##__VA_ARGS__})
+// There are 2 kinds of counters. `counter` is the general type of counter that are implemented
+// by striped_long_adder, which can achieve high performance while consuming less memory if
+// it's not updated very frequently. `counter2` uses concurrent_long_adder as the underlying
+// implementation. It has higher performance while consuming more memory if it's updated very
+// frequently. See also include/dsn/utility/long_adder.h for details.
```
Comment (acelyc111) [inline include/dsn/utility/metrics.h:76]:
```suggestion
// There are 2 kinds of counters.
// - `counter` is the general type of counter that are implemented by
//   striped_long_adder, which can achieve high performance while consuming
//   less memory if it's not updated very frequently.
// - `counter2` uses concurrent_long_adder as the underlying implementation.
//   It has higher performance while consuming more memory if it's updated very
//   frequently.
// See also include/dsn/utility/long_adder.h for details.
```

**Review by acelyc111 — APPROVED** [review]

File: src/utils/test/metrics_test.cpp [hunk 3]
```diff
@@ -313,11 +325,140 @@ TEST(metrics_test, gauge_double)
...
+    for (int64_t i = 0; i < num_threads; i++) {
+        threads.emplace_back([i, &runner]() { runner(i); });
+    }
+    for (auto &t : threads) {
+        t.join();
+    }
+}
+
+template <typename Adder>
+void run_counter_increment_by(::dsn::counter_ptr<Adder> &my_metric,
```
Comment (levy5307) [inline src/utils/test/metrics_test.cpp:347]: ::dsn::counter_ptr<Adder> &my_metric 
==>
dsn::counter_ptr<Adder> &my_metric

File: include/dsn/utility/metrics.h [hunk 4]
```diff
@@ -68,11 +69,27 @@
...
 #define METRIC_DEFINE_gauge_double(entity_type, name, unit, desc, ...)                             \
     ::dsn::gauge_prototype<double> METRIC_##name({#entity_type, #name, unit, desc, ##__VA_ARGS__})
+// There are 2 kinds of counters:
+// - `counter` is the general type of counter that is implemented by striped_long_adder, which can
+//   achieve high performance while consuming less memory if it's not updated very frequently.
+// - `concurrent_counter` uses concurrent_long_adder as the underlying implementation. It has
+//   higher performance while consuming more memory if it's updated very frequently.
+// See also include/dsn/utility/long_adder.h for details.
+#define METRIC_DEFINE_counter(entity_type, name, unit, desc, ...)                                  \
+    ::dsn::counter_prototype<::dsn::striped_long_adder> METRIC_##name(                             \
```
Comment (levy5307) [inline include/dsn/utility/metrics.h:79]: same

**Review by levy5307 — APPROVED** [review]

**Pull request merged**